flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorImpl.java

```diff
@@ -51,27 +66,40 @@ class ChannelStateWriteRequestExecutorImpl implements ChannelStateWriteRequestEx
     private final Thread thread;
     private volatile Exception thrown = null;
     private volatile boolean wasClosed = false;
-    private final String taskName;
+
+    private final Map<SubtaskID, Queue<ChannelStateWriteRequest>> unreadyQueues =
+            new ConcurrentHashMap<>();
+
+    private final JobID jobID;
+    private final Set<SubtaskID> subtasks;
+    private final AtomicBoolean isRegistering = new AtomicBoolean(true);
+    private final int numberOfSubtasksShareFile;
```

Comment:
Sorry, you mentioned it before. 

However, I encountered some difficulties during I refactor it:

- `BlockingDeque.take()` will  wait until an element becomes available. `Deque` is hard to achieve. `BlockingDeque` is easy to implement the `producer & consumer` model.
- The `loop` method access these fields continuously, do you have any suggestions to  refactor them using the lock?


```
private void loop() throws Exception {
    while (!wasClosed) {
        try {
            ChannelStateWriteRequest request = deque.take();
            // The executor will end the registration, when the start request comes.
            // Because the checkpoint can be started after all tasks are initiated.
            if (request instanceof CheckpointStartRequest && isRegistering()) {
                isRegistering.set(false);
            }
            dispatcher.dispatch(request);
        } catch (InterruptedException e) {
            if (!wasClosed) {
                LOG.debug(
                        "Channel state executor is interrupted while waiting for a request (continue waiting)",
                        e);
            } else {
                Thread.currentThread().interrupt();
            }
        }
    }
}
```

Sorry for not letting you know this before.

Reply:
> BlockingDeque.take() will wait until an element becomes available. Deque is hard to achieve. BlockingDeque is easy to implement the producer & consumer model.

That's a slight complexity, but should be easily solved via `lock.wait()` and `lock.notifyAll()` called in one or two places (`close()` and whenever we add anything to the current `dequeue`)? https://howtodoinjava.com/java/multi-threading/wait-notify-and-notifyall-methods/

The loop probably should look like this:

```
while (true) {
  synchronized (lock) {
    if (wasClosed) return; 
    (...)
    ChannelStateWriteRequest request = waitAndTakeUnsafe();
    (...)
  }
}

private ChannelStateWriteRequest waitAndTakeUnsafe() {
  ChannelStateWriteRequest request;
  while (true) {
    request = dequeue.pollFirst();
    if (request == null) {
      lock.wait();
    }
    else {
      return request;
    }
}
```

Reply:
Thanks for your suggestion, I have updated.

I didn't call the `lock.notifyAll()` during close due to we have called the `thread.interrupt()`. Please correct me if i'm wrong, thanks~

Reply:
Hmmm, I'm not sure. Maybe it's sufficient to have just the interrupt. On the other hand, does it hurt if we add the `lock.notifyAll()`?

Reply:
This is code about `lock.wait()`, if we add the `lock.notifyAll()`, the `lock.wait()` will return directly. And the `request` is still null, `lock.wait` will be blocked again. So the `lock.notifyAll()` doesn't work. That's why I think the `thread.interrupt()` is enough here.

```
    private ChannelStateWriteRequest waitAndTakeUnsafe() throws InterruptedException {
        ChannelStateWriteRequest request;
        while (true) {
            request = deque.pollFirst();
            if (request == null) {
                lock.wait();
            } else {
                return request;
            }
        }
    }
```

Reply:
Sorry, after address your next [comment](https://github.com/apache/flink/pull/20151#discussion_r1081119337), I found we can add the `lock.notifyAll()` here.

When the executor is closed, the while loop inside of `waitAndTakeUnsafe` will be finished, then `waitAndTakeUnsafe` will return `null`, and the `executor thread` can be finished.